### ISSUE TITLE
tf.nn.max_pool_with_argmax bug with padding

### ISSUE THREAD

**Issue by Panaetius** [tensorflow/tensorflow#10216]
There is a bug in the indices returned from `tf.nn.max_pool_with_argmax` when a padding is applied. The indices with be based on the shape supplied to `max_pool_with_argmax`, instead of the shape+padding.

Simple code example to reproduce:
```
import tensorflow as tf

def main():
    with tf.Session() as session:
        input = tf.get_variable('weights',
                                  shape=[1, 301, 201, 1],
                                  initializer=tf.truncated_normal_initializer(stddev=0.5, dtype=tf.float32),
                                  dtype=tf.float32)

        val, idx = tf.nn.max_pool_with_argmax(input, [1, 2, 2, 1], [1, 2, 2, 1], padding="SAME") #padding will turn dimensions to 302x202

        y1 = idx // 201
        x1 = idx % 201

        y2 = idx // 202
        x2 = idx % 202

        max_x1 = tf.reduce_max(x1)
        max_y1 = tf.reduce_max(y1)
        max_x2 = tf.reduce_max(x2)
        max_y2 = tf.reduce_max(y2)


        session.run(tf.global_variables_initializer())
        m_x1, m_y1, m_x2, m_y2 = session.run([max_x1, max_y1, max_x2, max_y2])

        print("%d, %d, %d, %d"%(m_x1, m_y1, m_x2, m_y2))

if __name__ == "__main__":
    main()
```

This prints `"200, 300, 201, 299"`. As you can see, the padding would increase the dimensions of the tensor to `302x202`, so the maximum `y` coordinate should be `301` and `x` should be `201`. But if we unravel the argmax indices with a width of `202`, we get a maximum `x` of `201`, but maximum `y` of only `299`. If we instead use `201` as width for unraveling (the unpadded width of the input tensor), we get `200` and `300`, respectively, which are the correct values for the unpadded input tensor.

So the `((b * height + y) * width + x) * channels + c` formula for `tf.nn.max_pool_with_argmax` uses the input tensor dimensions for `width`, not the input+padding dimensions.

This is relevant if you then use the indices to unpool/reverse the max_pool, since often you'll multiple the dimensions of the max_pool output by 2 to get the input dimensions, which would be off with a naive implementation like this.

When using this to implement an unpooling operation (for instance for SegNet), this will cause every line of the image to shift (by 1 pixel) if padding is applied to the width of an image,  basically slightly tilting an image. It's especially obvious with multiple argmax&unpool in succession.

It also means that, with zero-padding, if the whole tensor is negative values (in which case the zero-padding would be the highest value in the tensor), it won't return the coordinates of the padding. Basically, it doesn't actually add any padding to the input tensor, it just pretends it does to make dimensions line up, but doesn't consider the actual values/zeros in the padding. This might be intended behaviour, though it would strike me as odd given the usual understanding of padding. If so, the documentation should be changed to make it clear that this is the way the operation works. But I think this would be against the principle of least astonishment, since I assume most people would think that an op talking about padding would actually add padding values.

See also #2169 for a use-case for this with additional discussion

**Comment by jart** [comment]
@girving Since you participated a lot in the linked context thread, would you mind taking a look at this?

**Comment by girving** [comment]
I'm not sure whether this is intended behavior or not, but regardless I don't think we can change it.  It doesn't strike as clearly a bug, and changing it will break existing models and code in incredibly difficult to spot ways.

@Panaetius For your purposes, can you just add the padding-related offset?

Note that if an existing gradient routine uses the routine assuming the other behavior, and is therefore wrong, that's definitely a bug we should fix.  Similarly, if this routine fails to return negative indices in certain cases, I'd also consider that a bug which we should fix.  However, by itself the offset will probably have to remain.

**Comment by Panaetius** [comment]
@girving I'm currently not really affected, by supplying the width without padding to my unpool op, it's working fine.

If this isn't changed, then you should update the documentation of tf.nn.max_pool_with_argmax to clearly reflect this behaviour.

I can understand your reasoning, though I still regard it as a bug. Specifically, if there's [[-1, -2], [0, 0]] at the bottom of a tensor (the 0's being padding), this op would return the index of the -1, instead of one of the 0's, and that's not really what max_pooling should do. The way I see it right now, it's clearly a bug, but one without a lot of consequences (since it only affects the edges of an image), i.e. the example I just mentioned probably doesn't influence network performance, it might even be beneficial (not introducing arbitrary zeros). But it is conceptually a bug. 

That said, if the documentation is changed, it shouldn't make much of a difference. At least I can't think of a use-case where the behaviour can't be easily circumvented (like supplying the original dimensions in the unpool example), so I don't see it as a bug that would prohibit someone from doing what they want.

**Comment by michaelisard** [comment]
@girving would you take a stab at updating the docs or reassign?

**Comment by girving** [comment]
Hmm, pretty annoying that `max_pool_with_argmax` is GPU only. :/

**Comment by girving** [comment]
Fixing the documentation now.  I agree this is a conceptual bug, but I am too frightened to fix it since it could easily cause crashes.

**Comment by ScorpioCPH** [comment]
@girving @Panaetius 

I am sorry to comment on this closed issue, but I think `max_pool` has same behaviour with `max_pool_with_argmax `

Code:
```python
import tensorflow as tf
import numpy as np

M = np.array([
        [[1],[-1],[0]],
        [[-1],[2],[1]],
        [[0],[2],[-2]]
    ])

M = np.asarray(M, dtype='float32')
M = M.reshape(1, 3, 3, 1)

data = tf.placeholder("float32", shape=[1, 3, 3, 1])
pool = tf.nn.max_pool(data, [1, 2, 2, 1], [1, 2, 2, 1], padding='SAME')

print(pool.get_shape())

with tf.Session() as sess:
    pooled_M = sess.run(pool,feed_dict={data:M})
    print "pooled_M: \n", pooled_M
```

Output:
```shell
pooled_M: 
[[[[ 2.]
   [ 1.]]

  [[ 2.]
   [-2.]]]]
```

The last element should be `0` with `zero-padding` as `0` is bigger than `-2`.

WDYT?

**Comment by girving** [comment]
If they have the same bug, does that just mean the padding is with -inf?

**Comment by ScorpioCPH** [comment]
@girving thank you for the reply.

Cloud you give me more details? what is `-inf` meaning?

I think i can working on this issue with more details.

Firstly, i will add more unit tests for this ops, @girving WDYT?

**Comment by girving** [comment]
emmy:tensorflow% python3
    Python 3.6.1 (default, Apr 26 2017, 22:21:58) 
    [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.41)] on darwin
    Type "help", "copyright", "credits" or "license" for more information.
    >>> -inf
    -inf

**Comment by ScorpioCPH** [comment]
@girving Oh, i got it, but i don't think so.

Use `avg_pool` instead of `max_pool` of the previous code:

Code:
```python
import tensorflow as tf
import numpy as np

M = np.array([
        [[1],[-1],[0]],
        [[-1],[2],[1]],
        [[0],[2],[-2]]
    ])

M = np.asarray(M, dtype='float32')
M = M.reshape(1, 3, 3, 1)

data = tf.placeholder("float32", shape=[1, 3, 3, 1])
pool = tf.nn.avg_pool(data, [1, 2, 2, 1], [1, 2, 2, 1], padding='SAME')

print(pool.get_shape())

with tf.Session() as sess:
    pooled_M = sess.run(pool,feed_dict={data:M})
    print "pooled_M: \n", pooled_M
```

Output: 
```shell
pooled_M: 
[[[[ 0.25]
   [ 0.5 ]]

  [[ 1.  ]
   [-2.  ]]]]
```

It seems like `padding` values is just not involved in calculation.

**Comment by girving** [comment]
+ and max have different identity elements: 0 for + and -inf for max.

**Comment by ScorpioCPH** [comment]
@girving how about `avg_pool`?
